external/storm-kafka-new-consumer-api/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java

```diff
@@ -0,0 +1,457 @@
+                if ((currOffset = ackedMsg.offset()) == toCommitOffset || currOffset == toCommitOffset + 1) {    // found the next offset to commit
+                    toCommitMsgs.add(ackedMsg);
+                    toCommitMsg = ackedMsg;
+                    toCommitOffset = currOffset;
+                } else if (ackedMsg.offset() > toCommitOffset + 1) {    // offset found is not continuous to the offsets listed to go in the next commit, so stop search
+                    break;
+                } else {
+                    LOG.debug("Unexpected offset found [{]}. {}", ackedMsg.offset(), toString());
+                    break;
+                }
+            }
+
+            if (!toCommitMsgs.isEmpty()) {
+                offsetAndMetadata = new OffsetAndMetadata(toCommitOffset, toCommitMsg.getMetadata(Thread.currentThread()));
+                LOG.debug("Last offset to be committed in the next commit call: [{]}", offsetAndMetadata.offset());
+                LOG.trace(toString());
+            } else {
+                LOG.debug("No offsets found ready to commit");
+                LOG.trace(toString());
+            }
+            return offsetAndMetadata;
+        }
+
+        /**
+         * This method has side effects and should be called after findOffsetToCommit
+         */
+        public void updateAckedState(OffsetAndMetadata offsetAndMetadata) {
+            if (offsetAndMetadata != null) {
+                committedOffset = offsetAndMetadata.offset();
+                toCommitMsgs = new TreeSet<>(OFFSET_COMPARATOR);
+                ackedMsgs.removeAll(toCommitMsgs);
+            }
+        }
+
+        public boolean isEmpty() {
+            return ackedMsgs.isEmpty();
+        }
+
+        @Override
+        public String toString() {
+            return "OffsetEntry{" +
+                    "committedOffset=" + committedOffset +
+                    ", toCommitOffset=" + toCommitOffset +
+                    ", ackedMsgs=" + ackedMsgs +
+                    ", toCommitMsgs=" + toCommitMsgs +
+                    '}';
+        }
+    }
+
+    // =========== Consumer Rebalance Listener - On the same thread as the caller ===========
+
+    private class KafkaSpoutConsumerRebalanceListener implements ConsumerRebalanceListener {
+        @Override
+        public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
+            LOG.debug("Partitions revoked. [consumer-group={}, consumer={}, topic-partitions={}]",
+                    kafkaSpoutConfig.getConsumerGroupId(), kafkaConsumer, partitions);
+            commitAckedTuples();  // commit acked records,
+            copyEmittedTuplesToBlackList(); // all the tuples that are in traffic when the rebalance occurs will be added to black list to avoid duplication
+            clearFailed();   // remove all failed tuples form list to avoid duplication
+            clearEmittedTuples();    // clear emitted tuples
```

Comment:
Committing the acked tuples is done with a lock, but copying to the blacklist followed by clearing the others are not thread safe.  In fact the references to the maps are not volatile, so we could lose track of some tuples when this happens.

I can see that adding locking to the critical path could be costly, but I would rather do that then risk having state get in an odd place.